{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "\n",
    "        \n",
    "#Calculate percentage of each amino acid in the cell\n",
    "#1- Sum normalized ratios\n",
    "norm_sum = 0.\n",
    "for letter in alphabet:\n",
    "    value  = normalized_dict.get(letter)\n",
    "    norm_sum  = value + norm_sum\n",
    "    \n",
    "print(norm_sum)\n",
    "\n",
    "#2- Divide letter to norm_sum to get ratio of amino acid in the cell\n",
    "ratio_dict = {'A':0.,'C':0.,'D':0.,'E':0.,'F':0.,'G':0.,'H':0.,'I':0.,\n",
    "              'K':0.,'L':0.,'M':0.,'N':0.,'P':0.,'Q':0.,'R':0.,'S':0.,'T':0.,'V':0.,'W':0.,'Y':0.}\n",
    "\n",
    "#Constant for the amount of protein in the cell\n",
    "PROT_G = 5.9\n",
    "for letter in alphabet:\n",
    "    value = normalized_dict.get(letter)\n",
    "    ratio  = value/norm_sum\n",
    "    #Convert ratios to grams\n",
    "    converted_ratio = ratio*PROT_G    \n",
    "    ratio_dict[letter] = converted_ratio\n",
    "    \n",
    "    \n",
    "#3- Convert gram ratios to mmol/g Dry weight\n",
    "\n",
    "To verify that the normalized to grams to get to the total amount of protein \n",
    "(here everything is converted to grams instead of femto grams)\n",
    "new_value = 0.\n",
    "for letter in alphabet:\n",
    "    value = ratio_dict.get(letter)\n",
    "    new_value = new_value+value \n",
    "    \n",
    "print(new_value)\n",
    "\n",
    "letter_to_bigg = {'A':Mflorum.metabolites.ala__L_c,'C':Mflorum.metabolites.cys__L_c,'D':Mflorum.metabolites.asp__L_c,'E':Mflorum.metabolites.glu__L_c,'F':Mflorum.metabolites.phe__L_c,\n",
    "                  'G':Mflorum.metabolites.gly_c,'H':Mflorum.metabolites.his__L_c,'I':Mflorum.metabolites.ile__L_c,'K':Mflorum.metabolites.lys__L_c,'L':Mflorum.metabolites.leu__L_c,\n",
    "                  'M':Mflorum.metabolites.met__L_c,'N':Mflorum.metabolites.asn__L_c,'P':Mflorum.metabolites.pro__L_c,'Q':Mflorum.metabolites.gln__L_c,'R':Mflorum.metabolites.arg__L_c,\n",
    "                  'S':Mflorum.metabolites.ser__L_c,'T':Mflorum.metabolites.thr__L_c,'V':Mflorum.metabolites.val__L_c,'W':Mflorum.metabolites.trp__L_c,'Y':Mflorum.metabolites.tyr__L_c}\n",
    "CELL_WEIGHT = 9.7\n",
    "biomass_ratios = {'A':0.,'C':0.,'D':0.,'E':0.,'F':0.,'G':0.,'H':0.,'I':0.,\n",
    "                  'K':0.,'L':0.,'M':0.,'N':0.,'P':0.,'Q':0.,'R':0.,'S':0.,'T':0.,'V':0.,'W':0.,'Y':0.}\n",
    "#Get number of moles from number of grams\n",
    "for letter in alphabet:\n",
    "    metab = letter_to_bigg.get(letter)\n",
    "    mol_weight = metab.formula_weight\n",
    "    grams = ratio_dict.get(letter)\n",
    "    mmols_per_cell = (grams/mol_weight)*1000\n",
    "    mmols_per_gDW = mmols_per_cell/CELL_WEIGHT\n",
    "    biomass_ratios[letter] = mmols_per_gDW\n",
    "    \n",
    "\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "Update for protein\n",
    "#Change the ratios based on their appearance in the reactants and products of biomass eq\n",
    "'''\n",
    "Algorythm should go like this:\n",
    "- Find metabolites from BIOMASS (stack them in a list with their coefficients)\n",
    "- Remove the metabolites from the BIOMASS equation\n",
    "- Add them with the new coefficients\n",
    "'''\n",
    "#1- Find metabolites in BIOMASS\n",
    "reactants_biomass, biomass_coefficients, reactant_removal =[],[],[]\n",
    "for letter in alphabet:\n",
    "    metab = letter_to_bigg.get(letter)\n",
    "    coefficient = biomass_ratios.get(letter)\n",
    "    for r in Mflorum.reactions.BIOMASS.reactants:\n",
    "        if metab.id[0:3] in r.id:\n",
    "            reactants_biomass.append(r)\n",
    "            neg_coefficient = -coefficient\n",
    "            biomass_coefficients.append(neg_coefficient)\n",
    "            reactant_removal.append(1)\n",
    "\n",
    "new_reactants_coefficients = dict(zip(reactants_biomass,biomass_coefficients))\n",
    "remove_reactants = dict(zip(reactants_biomass,reactant_removal))\n",
    "\n",
    "products_biomass, biomass_coefficients, product_removal =[],[],[]\n",
    "for letter in alphabet:\n",
    "    metab = letter_to_bigg.get(letter)\n",
    "    coefficient = biomass_ratios.get(letter)\n",
    "    for r in Mflorum.reactions.BIOMASS.products:\n",
    "        if metab.id[0:3] in r.id:\n",
    "            products_biomass.append(r)\n",
    "            biomass_coefficients.append(coefficient)\n",
    "            product_removal.append(-1)\n",
    "            \n",
    "new_products_coefficients = dict(zip(amino_acids_biomass,biomass_coefficients))\n",
    "remove_products = dict(zip(products_biomass,product_removal))\n",
    "    \n",
    "remove_reactants\n",
    "\n",
    "#2- Remove metabolites from BIOMASS equation\n",
    "Mflorum.reactions.BIOMASS.add_metabolites(remove_reactants)\n",
    "print(Mflorum.reactions.BIOMASS.reactants)\n",
    "\n",
    "#3- Add reactants and products to biomass with their calculated ratios\n",
    "Mflorum.reactions.BIOMASS.add_metabolites(new_reactants_coefficients)\n",
    "print(Mflorum.reactions.BIOMASS.reactants)\n",
    "\n",
    "#3- Add reactants and products to biomass with their calculated ratios\n",
    "Mflorum.reactions.BIOMASS.add_metabolites(new_products_coefficients)\n",
    "print(Mflorum.reactions.BIOMASS.reaction)\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Generate an example input file from proteomics data\n",
    "#The format is 2 columns, gene_id:abundance (could be csv, dataframe or dictionary)\n",
    "\n",
    "#Obtain the total amount of protein from proteomic data\n",
    "proteomics = pd.read_table('/home/jean-christophe/Documents/Maitrise_UCSD/Experimental validation/Proteomics/Genes_in_model_expression_list_summary.txt')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Utilitaries():\n",
    "    \n",
    "    def chopStr(text):\n",
    "        return (text[2:-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n"
     ]
    }
   ],
   "source": [
    "text = '01234'\n",
    "Utilitaries.chopStr(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Protein(Utilitaries,Update):\n",
    "    import pandas as pd\n",
    "    \n",
    "    #Objects\n",
    "    #none so far\n",
    "    #Methods\n",
    "    def get_coefficients(path_to_genbank,path_to_model,proteomics,CELL_WEIGHT=0.3, PROTEIN_RATIO=0.6):\n",
    "        \n",
    "        AMINO_ACIDS = ['A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y']\n",
    "        \n",
    "        def import_genome(genbank):\n",
    "            from Bio import SeqIO\n",
    "            return SeqIO.read(genbank,'genbank')\n",
    "        \n",
    "        def import_model(path_to_model):\n",
    "            import cobra\n",
    "            extension = path_to_model.split('.')[-1]\n",
    "            if extension == 'json':\n",
    "                return cobra.io.load_json_model(path_to_model)\n",
    "            elif extension == 'xml':\n",
    "                return cobra.io.read_sbml_model(path_to_model)\n",
    "            \n",
    "        def get_protein_sequence(genome_record):\n",
    "            #Get the sequence of each protein, locus tag and prot id from genbank file\n",
    "            seq_list,protID_list,locus_list = [],[],[]\n",
    "            for element in genome_record.features:\n",
    "                if element.type == 'CDS':\n",
    "                    locus_list.append(element.qualifiers['locus_tag'][0])\n",
    "                    protID_list.append(element.qualifiers['protein_id'][0])\n",
    "                    seq_list.append(element.qualifiers['translation'][0])\n",
    "\n",
    "            return pd.DataFrame({'Locus_tag':locus_list, 'Protein_ID':protID_list, 'Sequence':seq_list})\n",
    "        \n",
    "        def get_aa_composition(df1):\n",
    "        #For each protein find the amino acid composition\n",
    "        #Outputs a dictionnary of dictionnaries where:\n",
    "        #Keys = locus_tag\n",
    "        #Values = A dictionary for each amino acid\n",
    "            #This dictionary contains:\n",
    "            #Keys = amino acid by letter code\n",
    "            #Values = the occurence of that amino acid\n",
    "            list_of_dict =[]\n",
    "            for i, row in df1.iterrows():\n",
    "                list_of_occurences = []\n",
    "                #Get the occurence for each letter\n",
    "                for letter in AMINO_ACIDS:\n",
    "                    protein_sequence = row.Sequence\n",
    "                    occurence_of_letter = protein_sequence.count(letter)\n",
    "                    list_of_occurences.append(occurence_of_letter)\n",
    "                #Generate dictionary of occurences for a given gene\n",
    "                dict_of_occurences = dict(zip(AMINO_ACIDS,list_of_occurences))\n",
    "                #Generate dict for each gene\n",
    "                dict_per_locus = {row.Locus_tag:dict_of_occurences}\n",
    "                #Store the amount of each amino acid per gene in a list\n",
    "                list_of_dict.append(dict_per_locus)\n",
    "                \n",
    "            return list_of_dict\n",
    "\n",
    "        def normalize_aa_composition(list_of_dict):\n",
    "            #Normalize the value of each amino acid per protein following proteomic data\n",
    "            normalized_dict = {'A':0.,'C':0.,'D':0.,'E':0.,'F':0.,'G':0.,'H':0.,'I':0.,\n",
    "                               'K':0.,'L':0.,'M':0.,'N':0.,'P':0.,'Q':0.,'R':0.,'S':0.,'T':0.,'V':0.,'W':0.,'Y':0.}\n",
    "            for d in list_of_dict:\n",
    "                #Get the coefficient from proteomics\n",
    "                text = str(list(d.keys())[0])\n",
    "                coeff = coeff_dict.get(text)\n",
    "                #If no protein abundance coefficient is 0.\n",
    "                try:\n",
    "                    coeff_number = float(coeff)\n",
    "                except:\n",
    "                    coeff_number = 0.\n",
    "                    \n",
    "                #Multiply each amino acid by the coefficient\n",
    "                amino_acids = list(d.values())\n",
    "                for letter in AMINO_ACIDS:\n",
    "                    value = float(amino_acids[0].get(letter))\n",
    "                    #Update the normalized value\n",
    "                    normalized_value = value*coeff_number\n",
    "                    new_value = normalized_dict.get(letter) + normalized_value\n",
    "                    normalized_dict[letter] = new_value\n",
    "                    \n",
    "            return normalized_dict\n",
    "         \n",
    "        #Operations\n",
    "        #1- Parse the genome, extract protein sequence, count and store amino acid composition of each protein\n",
    "        genome_record = import_genome(path_to_genbank)\n",
    "        df1 = get_protein_sequence(genome_record)\n",
    "        list_of_dict = get_aa_composition(df1)\n",
    "        normalized_dict = normalize_aa_composition(list_of_dict)\n",
    "        \n",
    "        def get_norm_sum(normalized_dict):\n",
    "            #1- Sum normalized ratios\n",
    "            norm_sum = 0.\n",
    "            for letter in AMINO_ACIDS:\n",
    "                value  = normalized_dict.get(letter)\n",
    "                norm_sum  = value + norm_sum\n",
    "\n",
    "            return norm_sum\n",
    "\n",
    "        def get_ratio(normalized_dict,norm_sum,PROTEIN_RATIO):\n",
    "            #2- Divide letter to norm_sum to get ratio of each amino acid in the cell \n",
    "            # based on proteomic data\n",
    "            ratio_dict = {'A':0.,'C':0.,'D':0.,'E':0.,'F':0.,'G':0.,'H':0.,'I':0.,\n",
    "                          'K':0.,'L':0.,'M':0.,'N':0.,'P':0.,'Q':0.,'R':0.,'S':0.,'T':0.,'V':0.,'W':0.,'Y':0.}\n",
    "\n",
    "            #Constant for the amount of protein in the cell\n",
    "            PROTEIN_WEIGHT = CELL_WEIGHT*PROTEIN_RATIO\n",
    "            for letter in AMINO_ACIDS:\n",
    "                value = normalized_dict.get(letter)\n",
    "                ratio  = value/norm_sum\n",
    "                #Convert ratios to grams\n",
    "                converted_ratio = ratio*PROTEIN_WEIGHT    \n",
    "                ratio_dict[letter] = converted_ratio\n",
    "                \n",
    "            return ratio_dict\n",
    "    \n",
    "        def convert_to_coefficient(path_to_model,CELL_WEIGHT):\n",
    "            model = import_model(path_to_model)\n",
    "            #3- Convert gram ratios to mmol/g Dry weight\n",
    "            '''\n",
    "            To verify that the normalized to grams to get to the total amount of protein \n",
    "            (here everything is converted to grams instead of femto grams)\n",
    "            '''\n",
    "            letter_to_bigg = {'A':model.metabolites.ala__L_c,'C':model.metabolites.cys__L_c,'D':model.metabolites.asp__L_c,'E':model.metabolites.glu__L_c,'F':model.metabolites.phe__L_c,\n",
    "                              'G':model.metabolites.gly_c,'H':model.metabolites.his__L_c,'I':model.metabolites.ile__L_c,'K':model.metabolites.lys__L_c,'L':model.metabolites.leu__L_c,\n",
    "                              'M':model.metabolites.met__L_c,'N':model.metabolites.asn__L_c,'P':model.metabolites.pro__L_c,'Q':model.metabolites.gln__L_c,'R':model.metabolites.arg__L_c,\n",
    "                              'S':model.metabolites.ser__L_c,'T':model.metabolites.thr__L_c,'V':model.metabolites.val__L_c,'W':model.metabolites.trp__L_c,'Y':model.metabolites.tyr__L_c}\n",
    "\n",
    "            biomass_coefficients = {'A':0.,'C':0.,'D':0.,'E':0.,'F':0.,'G':0.,'H':0.,'I':0.,\n",
    "                              'K':0.,'L':0.,'M':0.,'N':0.,'P':0.,'Q':0.,'R':0.,'S':0.,'T':0.,'V':0.,'W':0.,'Y':0.}\n",
    "            #Get number of moles from number of grams\n",
    "            for letter in AMINO_ACIDS:\n",
    "                metab = letter_to_bigg.get(letter)\n",
    "                mol_weight = metab.formula_weight\n",
    "                grams = ratio_dict.get(letter)\n",
    "                mmols_per_cell = (grams/mol_weight)*1000\n",
    "                mmols_per_gDW = mmols_per_cell/CELL_WEIGHT\n",
    "                biomass_coefficients[letter] = mmols_per_gDW\n",
    "            \n",
    "            return biomass_coefficients\n",
    "\n",
    "        #2- Get coefficients from experimental proteomics data\n",
    "        #Proteomics data should come in a 2 columns standard format gene_id:abundance\n",
    "        norm_sum = get_norm_sum(normalized_dict)\n",
    "        ratio_dict = get_ratio(normalized_dict,norm_sum,PROTEIN_RATIO)\n",
    "        biomass_coefficients = convert_to_coefficient(path_to_model,CELL_WEIGHT)\n",
    "        print(biomass_coefficients)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'R': 0.12698594044292982, 'M': 0.10039334382981341, 'K': 0.4228955466570605, 'P': 0.14989571771940108, 'E': 0.32360384989600366, 'H': 0.05279661803935423, 'Q': 0.12769799109529054, 'V': 0.36217751226096007, 'Y': 0.09542866936675634, 'W': 0.01970009517084819, 'T': 0.2918719442117945, 'D': 0.24063727925750836, 'A': 0.5159110997956983, 'L': 0.3701278253832807, 'C': 0.024764164647747312, 'G': 0.514084868214509, 'F': 0.13775204382857809, 'N': 0.26425478270623226, 'I': 0.38866814279916834, 'S': 0.32422059315282087}\n"
     ]
    }
   ],
   "source": [
    "path_to_genbank = '/Users/JC/Documents/Maitrise_UCSD/GenBankFiles/Mflorum.gbff'\n",
    "path_to_model = '/Users/JC/Documents/Maitrise_UCSD/Functionnal_testing/finalMeso1.json'\n",
    "Protein.get_coefficients(path_to_genbank,path_to_model,proteomics,CELL_WEIGHT=0.3,PROTEIN_RATIO=0.6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/JC/Documents/Maitrise_UCSD/Functionnal_testing/biomass'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "IOError",
     "evalue": "File /Users/JC/Documents/Maitrise_UCSD/Experimental_validation/Proteomics/Uniprot_conversion.csv does not exist",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIOError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-be33a135b88e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0muniprot_conversion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/Users/JC/Documents/Maitrise_UCSD/Experimental_validation/Proteomics/Uniprot_conversion.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mpeptide_shaker\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/Users/JC/Documents/Maitrise_UCSD/Experimental_validation/Proteomics/20160723_2D_SRodrigue_ALL/reports/20160723_2D_SRodrigue_ALL_sample_1_Default_Hierarchical_Report.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, escapechar, comment, encoding, dialect, tupleize_cols, error_bad_lines, warn_bad_lines, skipfooter, skip_footer, doublequote, delim_whitespace, as_recarray, compact_ints, use_unsigned, low_memory, buffer_lines, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    644\u001b[0m                     skip_blank_lines=skip_blank_lines)\n\u001b[1;32m    645\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 646\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    647\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    648\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    387\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    388\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 389\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    390\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    391\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mnrows\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mchunksize\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    728\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'has_index_names'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    729\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 730\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    731\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    732\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m    921\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'c'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    922\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'c'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 923\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    924\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    925\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'python'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/lib/python2.7/dist-packages/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1388\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'allow_leading_cols'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1389\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1390\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_parser\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1391\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1392\u001b[0m         \u001b[0;31m# XXX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader.__cinit__ (pandas/parser.c:4184)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/parser.pyx\u001b[0m in \u001b[0;36mpandas.parser.TextReader._setup_parser_source (pandas/parser.c:8449)\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mIOError\u001b[0m: File /Users/JC/Documents/Maitrise_UCSD/Experimental_validation/Proteomics/Uniprot_conversion.csv does not exist"
     ]
    }
   ],
   "source": [
    "#Generate proteomics file to be manipulated by package\n",
    "import pandas as pd\n",
    "\n",
    "uniprot_conversion = pd.read_table('/Users/JC/Documents/Maitrise_UCSD/Experimental_validation/Proteomics/Uniprot_conversion.csv')\n",
    "peptide_shaker = pd.read_csv('/Users/JC/Documents/Maitrise_UCSD/Experimental_validation/Proteomics/20160723_2D_SRodrigue_ALL/reports/20160723_2D_SRodrigue_ALL_sample_1_Default_Hierarchical_Report.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Normalize by the abundance of protein given by proteomic data\n",
    "coefficients, accessions = [], []\n",
    "for i,row in peptide_shaker.iterrows():\n",
    "    n = str(row['Unnamed: 0'])\n",
    "    if '.' not in n and n != 'nan':\n",
    "        coefficients.append(row['Spectrum Counting NSAF'])\n",
    "        accessions.append(row['Main Accession'])  \n",
    "        \n",
    "coeff_accession = pd.DataFrame({'NSAF': coefficients, 'Accession': accessions})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "accessions2, mfl_ids = [],[]\n",
    "for i, row in uniprot_conversion.iterrows():\n",
    "    accessions2.append(row['yourlist:M20170121A7434721E10EE6586998A056CCD0537E674960X'])\n",
    "    mfl_ids.append(row['Gene names'])\n",
    "\n",
    "#Generate a conversion dict for protein accession numbers to locus tags\n",
    "import re\n",
    "new_mfl = []\n",
    "for element in mfl_ids:\n",
    "    new_mfl.append('Mfl'+ re.findall('(?<=Mfl).+', element)[0])\n",
    "    \n",
    "    \n",
    "conv_dict = dict(zip(accessions2,new_mfl))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mfl172</td>\n",
       "      <td>0.0716267146123262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mfl392</td>\n",
       "      <td>0.208204555050382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mfl052</td>\n",
       "      <td>0.0168350168350168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mfl306</td>\n",
       "      <td>0.132590145884944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mfl675</td>\n",
       "      <td>0.0341816078658184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Mfl091</td>\n",
       "      <td>0.0772421755594832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Mfl655</td>\n",
       "      <td>0.0523598820058997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Mfl201</td>\n",
       "      <td>2.0660815173527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Mfl613</td>\n",
       "      <td>0.0168300367164707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Mfl341</td>\n",
       "      <td>0.115384615384615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Mfl044</td>\n",
       "      <td>0.0388929987031253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Mfl438</td>\n",
       "      <td>0.00501672240802676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Mfl305</td>\n",
       "      <td>0.0233418826044726</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Mfl441</td>\n",
       "      <td>0.381649831649832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Mfl117</td>\n",
       "      <td>0.0406462585034014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Mfl025</td>\n",
       "      <td>0.0599472990777339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Mfl023</td>\n",
       "      <td>0.118136551249047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Mfl662</td>\n",
       "      <td>0.0324858757062147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Mfl236</td>\n",
       "      <td>0.0175257731958763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Mfl240</td>\n",
       "      <td>0.0920722169986876</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Mfl048</td>\n",
       "      <td>0.208279006686486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Mfl583</td>\n",
       "      <td>0.00108225108225108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Mfl374</td>\n",
       "      <td>0.0421348314606742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Mfl199</td>\n",
       "      <td>0.0645186089630534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Mfl095</td>\n",
       "      <td>0.0928343586238323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Mfl129</td>\n",
       "      <td>0.156873919684619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Mfl394</td>\n",
       "      <td>0.0224719101123595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Mfl349</td>\n",
       "      <td>0.0347407742439699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Mfl124</td>\n",
       "      <td>0.119016515319888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Mfl128</td>\n",
       "      <td>0.432834836294659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>Mfl235</td>\n",
       "      <td>0.0341409078251183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>Mfl513</td>\n",
       "      <td>0.0557040082219939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>Mfl602</td>\n",
       "      <td>0.716764930757468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>Mfl176</td>\n",
       "      <td>0.122904165186044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>Mfl551</td>\n",
       "      <td>0.014336917562724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>Mfl028</td>\n",
       "      <td>0.0528846153846154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>Mfl514</td>\n",
       "      <td>0.0849592366129331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>Mfl597</td>\n",
       "      <td>0.0509682745896966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>Mfl493</td>\n",
       "      <td>0.431999283341566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td>Mfl516</td>\n",
       "      <td>0.0617529705871052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>Mfl379</td>\n",
       "      <td>0.0402909354147744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>Mfl315</td>\n",
       "      <td>0.00922909880564604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>Mfl006</td>\n",
       "      <td>0.0266089291410208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>Mfl262</td>\n",
       "      <td>0.0051153422018562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>Mfl125</td>\n",
       "      <td>0.337964370145221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>Mfl635</td>\n",
       "      <td>0.0868104667609619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>Mfl105</td>\n",
       "      <td>0.141977691977692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>Mfl418</td>\n",
       "      <td>0.0350969669793219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>Mfl645</td>\n",
       "      <td>0.0223880597014925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>Mfl402</td>\n",
       "      <td>0.0111111111111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>482</th>\n",
       "      <td>Mfl587</td>\n",
       "      <td>0.0467131850874324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>483</th>\n",
       "      <td>Mfl674</td>\n",
       "      <td>0.0159722222222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>Mfl406</td>\n",
       "      <td>0.0808080808080808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>Mfl395</td>\n",
       "      <td>0.0321111111111111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>Mfl119</td>\n",
       "      <td>0.0240753216503794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>Mfl143</td>\n",
       "      <td>0.0120232172470978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>Mfl050</td>\n",
       "      <td>0.0154320987654321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>Mfl113</td>\n",
       "      <td>0.199385515269999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>490</th>\n",
       "      <td>Mfl481</td>\n",
       "      <td>0.0242914979757085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>Mfl252</td>\n",
       "      <td>0.0219164118246687</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>492 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_id            abundance\n",
       "0    Mfl172   0.0716267146123262\n",
       "1    Mfl392    0.208204555050382\n",
       "2    Mfl052   0.0168350168350168\n",
       "3    Mfl306    0.132590145884944\n",
       "4    Mfl675   0.0341816078658184\n",
       "5    Mfl091   0.0772421755594832\n",
       "6    Mfl655   0.0523598820058997\n",
       "7    Mfl201      2.0660815173527\n",
       "8    Mfl613   0.0168300367164707\n",
       "9    Mfl341    0.115384615384615\n",
       "10   Mfl044   0.0388929987031253\n",
       "11   Mfl438  0.00501672240802676\n",
       "12   Mfl305   0.0233418826044726\n",
       "13   Mfl441    0.381649831649832\n",
       "14   Mfl117   0.0406462585034014\n",
       "15   Mfl025   0.0599472990777339\n",
       "16   Mfl023    0.118136551249047\n",
       "17   Mfl662   0.0324858757062147\n",
       "18   Mfl236   0.0175257731958763\n",
       "19   Mfl240   0.0920722169986876\n",
       "20   Mfl048    0.208279006686486\n",
       "21   Mfl583  0.00108225108225108\n",
       "22   Mfl374   0.0421348314606742\n",
       "23   Mfl199   0.0645186089630534\n",
       "24   Mfl095   0.0928343586238323\n",
       "25   Mfl129    0.156873919684619\n",
       "26   Mfl394   0.0224719101123595\n",
       "27   Mfl349   0.0347407742439699\n",
       "28   Mfl124    0.119016515319888\n",
       "29   Mfl128    0.432834836294659\n",
       "..      ...                  ...\n",
       "462  Mfl235   0.0341409078251183\n",
       "463  Mfl513   0.0557040082219939\n",
       "464  Mfl602    0.716764930757468\n",
       "465  Mfl176    0.122904165186044\n",
       "466  Mfl551    0.014336917562724\n",
       "467  Mfl028   0.0528846153846154\n",
       "468  Mfl514   0.0849592366129331\n",
       "469  Mfl597   0.0509682745896966\n",
       "470  Mfl493    0.431999283341566\n",
       "471  Mfl516   0.0617529705871052\n",
       "472  Mfl379   0.0402909354147744\n",
       "473  Mfl315  0.00922909880564604\n",
       "474  Mfl006   0.0266089291410208\n",
       "475  Mfl262   0.0051153422018562\n",
       "476  Mfl125    0.337964370145221\n",
       "477  Mfl635   0.0868104667609619\n",
       "478  Mfl105    0.141977691977692\n",
       "479  Mfl418   0.0350969669793219\n",
       "480  Mfl645   0.0223880597014925\n",
       "481  Mfl402   0.0111111111111111\n",
       "482  Mfl587   0.0467131850874324\n",
       "483  Mfl674   0.0159722222222222\n",
       "484  Mfl406   0.0808080808080808\n",
       "485  Mfl395   0.0321111111111111\n",
       "486  Mfl119   0.0240753216503794\n",
       "487  Mfl143   0.0120232172470978\n",
       "488  Mfl050   0.0154320987654321\n",
       "489  Mfl113    0.199385515269999\n",
       "490  Mfl481   0.0242914979757085\n",
       "491  Mfl252   0.0219164118246687\n",
       "\n",
       "[492 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accession_mfl = []\n",
    "for i, row in coeff_accession.iterrows():\n",
    "    accession_mfl.append(conv_dict.get(row['Accession']))\n",
    "    \n",
    "coeff_dict = dict(zip(accession_mfl, coefficients))\n",
    "proteomics = pd.DataFrame({'gene_id':[k for k in coeff_dict.keys()], \n",
    "                           'abundance':[v for v in coeff_dict.values()]}\n",
    "                          ,columns=['gene_id','abundance'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
