

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>API &mdash; biomass 0.0.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="biomass 0.0.1 documentation" href="index.html"/>
        <link rel="next" title="License" href="license.html"/>
        <link rel="prev" title="NGAM" href="NGAM.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> biomass
          

          
          </a>

          
            
            
              <div class="version">
                0.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="DNA.html">DNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="RNA.html">RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="Protein.html">Protein</a></li>
<li class="toctree-l1"><a class="reference internal" href="Lipids.html">Lipids</a></li>
<li class="toctree-l1"><a class="reference internal" href="Metabolites.html">Metabolites</a></li>
<li class="toctree-l1"><a class="reference internal" href="GAM.html">GAM</a></li>
<li class="toctree-l1"><a class="reference internal" href="NGAM.html">NGAM</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#dna">DNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rna">RNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#protein">Protein</a></li>
<li class="toctree-l2"><a class="reference internal" href="#lipid">Lipid</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gam">GAM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="help.html">Help</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">biomass</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>API</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/API.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-DNA">
<span id="api"></span><h1>API<a class="headerlink" href="#module-DNA" title="Permalink to this headline">¶</a></h1>
<div class="section" id="dna">
<h2>DNA<a class="headerlink" href="#dna" title="Permalink to this headline">¶</a></h2>
<p>This module generates BOFsc for the 4 bases of DNA (dATP, dTTP, dCTP and dGTP)</p>
<dl class="function">
<dt id="DNA.get_coefficients">
<code class="descclassname">DNA.</code><code class="descname">get_coefficients</code><span class="sig-paren">(</span><em>path_to_fasta</em>, <em>path_to_model</em>, <em>CELL_WEIGHT=280</em>, <em>DNA_RATIO=0.031</em><span class="sig-paren">)</span><a class="headerlink" href="#DNA.get_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary of metabolite:coefficients for the 4 DNA bases from the organism’s
DNA fasta file and the weight percentage of DNA in the cell.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path_to_fasta</strong> – a path to the DNA fasta file of the organism, format should be compatible with BioPython SeqIO</li>
<li><strong>path_to_model</strong> – a path to the model, format supported are json and xml</li>
<li><strong>CELL_WEIGHT</strong> – experimentally measured cell weight in femtograms, float</li>
<li><strong>DNA_RATIO</strong> – the ratio of DNA in the entire cell</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary of metabolites and coefficients</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="DNA.update_biomass_coefficients">
<code class="descclassname">DNA.</code><code class="descname">update_biomass_coefficients</code><span class="sig-paren">(</span><em>dict_of_coefficients</em>, <em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#DNA.update_biomass_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the biomass coefficients given the input metabolite:coefficient dictionary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>dict_of_coefficients</strong> – dictionary of metabolites and coefficients</li>
<li><strong>model</strong> – model to update</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The biomass objective function is updated.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="DNA._import_genome">
<code class="descclassname">DNA.</code><code class="descname">_import_genome</code><span class="sig-paren">(</span><em>fasta</em><span class="sig-paren">)</span><a class="headerlink" href="#DNA._import_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function utilizes BioPython to read the fasta file provided by the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fasta</strong> – the path to the fasta file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the genome sequence to use for calculation</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="DNA._import_model">
<code class="descclassname">DNA.</code><code class="descname">_import_model</code><span class="sig-paren">(</span><em>path_to_model</em><span class="sig-paren">)</span><a class="headerlink" href="#DNA._import_model" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the model in either .json or .xml format</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_model</strong> – the path to the model. Supported format are .json and .xml</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">model</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-RNA"></span><div class="section" id="rna">
<h2>RNA<a class="headerlink" href="#rna" title="Permalink to this headline">¶</a></h2>
<p>This module generates BOFsc for the 4 bases of RNA (ATP, UTP, CTP and GTP)</p>
<dl class="function">
<dt id="RNA.get_coefficients">
<code class="descclassname">RNA.</code><code class="descname">get_coefficients</code><span class="sig-paren">(</span><em>path_to_genbank</em>, <em>path_to_model</em>, <em>path_to_transcriptomic</em>, <em>CELL_WEIGHT=280</em>, <em>TOTAL_RNA_RATIO=0.205</em>, <em>rRNA_RATIO=0.8</em>, <em>tRNA_RATIO=0.1</em>, <em>mRNA_RATIO=0.1</em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.get_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary of metabolite:coefficients for the 4 RNA bases from the organism’s
GenBank annotated file, total RNA weight percentage, transcriptomic. Alternately, ribosomal,
transfer and messenger RNA relative abundances can be incorporated otherwise the default 80% rRNA, 10% tRNA and
10% mRNA are used.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path_to_genbank</strong> – a path to the GenBank annotation file of the organism, format should be compatible with BioPython SeqIO</li>
<li><strong>path_to_model</strong> – a path to the model, format supported are json and xml</li>
<li><strong>proteomics</strong> – a two column pandas dataframe (gene_id, abundance)</li>
<li><strong>CELL_WEIGHT</strong> – experimentally measured cell weight in femtograms, float</li>
<li><strong>RNA_RATIO</strong> – the ratio of DNA in the entire cell</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary of metabolites and coefficients</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="RNA.update_biomass_coefficients">
<code class="descclassname">RNA.</code><code class="descname">update_biomass_coefficients</code><span class="sig-paren">(</span><em>dict_of_coefficients</em>, <em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#RNA.update_biomass_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the biomass coefficients given the input metabolite:coefficient dictionary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>dict_of_coefficients</strong> – dictionary of metabolites and coefficients</li>
<li><strong>model</strong> – model to update</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The biomass objective function is updated.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="RNA._import_genome">
<code class="descclassname">RNA.</code><code class="descname">_import_genome</code><span class="sig-paren">(</span><em>path_to_genbank</em><span class="sig-paren">)</span><a class="headerlink" href="#RNA._import_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function utilizes BioPython to read the GenBank annotation file provided by the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_genbank</strong> – the path GenBank annotation file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the genome record</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="RNA._import_model">
<code class="descclassname">RNA.</code><code class="descname">_import_model</code><span class="sig-paren">(</span><em>path_to_model</em><span class="sig-paren">)</span><a class="headerlink" href="#RNA._import_model" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the model in either .json or .xml format</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_model</strong> – the path to the model. Supported format are .json and .xml</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">model</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="RNA._import_transcriptomic">
<code class="descclassname">RNA.</code><code class="descname">_import_transcriptomic</code><span class="sig-paren">(</span><em>path_to_transcriptomic</em><span class="sig-paren">)</span><a class="headerlink" href="#RNA._import_transcriptomic" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the 2 column .csv file containing the relative abundances of each gene.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_transcriptomic</strong> – the path to to 2 column .csv file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">pandas DataFrame containing gene identifiers and relative abundances</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-Protein"></span><div class="section" id="protein">
<h2>Protein<a class="headerlink" href="#protein" title="Permalink to this headline">¶</a></h2>
<p>This module generates BOFsc for the 20 amino acids contained in proteins.</p>
<dl class="function">
<dt id="Protein.get_coefficients">
<code class="descclassname">Protein.</code><code class="descname">get_coefficients</code><span class="sig-paren">(</span><em>path_to_genbank</em>, <em>path_to_model</em>, <em>path_to_proteomic</em>, <em>CELL_WEIGHT=280</em>, <em>PROTEIN_RATIO=0.55</em><span class="sig-paren">)</span><a class="headerlink" href="#Protein.get_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary of metabolite:coefficients for the 20 amino acids contained in proteins from the organism’s
GenBank annotated file, total Protein weight percentage and proteomic data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path_to_genbank</strong> – a path to the GenBank annotation file of the organism, format should be compatible with BioPython SeqIO</li>
<li><strong>path_to_model</strong> – a path to the model, format supported are json and xml</li>
<li><strong>path_to_proteomic</strong> – a two column pandas dataframe (gene_id, abundance)</li>
<li><strong>CELL_WEIGHT</strong> – experimentally measured cell weight in femtograms, float</li>
<li><strong>PROTEIN_RATIO</strong> – the ratio of DNA in the entire cell</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary of metabolites and coefficients</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Protein.update_biomass_coefficients">
<code class="descclassname">Protein.</code><code class="descname">update_biomass_coefficients</code><span class="sig-paren">(</span><em>dict_of_coefficients</em>, <em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#Protein.update_biomass_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the biomass coefficients given the input metabolite:coefficient dictionary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>dict_of_coefficients</strong> – dictionary of metabolites and coefficients</li>
<li><strong>model</strong> – model to update</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The biomass objective function is updated.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Protein._import_genome">
<code class="descclassname">Protein.</code><code class="descname">_import_genome</code><span class="sig-paren">(</span><em>path_to_genbank</em><span class="sig-paren">)</span><a class="headerlink" href="#Protein._import_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function utilizes BioPython to read the GenBank annotation file provided by the user.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_genbank</strong> – the path GenBank annotation file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the genome record</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Protein._import_model">
<code class="descclassname">Protein.</code><code class="descname">_import_model</code><span class="sig-paren">(</span><em>path_to_model</em><span class="sig-paren">)</span><a class="headerlink" href="#Protein._import_model" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the model in either .json or .xml format</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_model</strong> – the path to the model. Supported format are .json and .xml</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">model</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Protein._import_proteomic">
<code class="descclassname">Protein.</code><code class="descname">_import_proteomic</code><span class="sig-paren">(</span><em>path_to_proteomic</em><span class="sig-paren">)</span><a class="headerlink" href="#Protein._import_proteomic" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the 2 column .csv file containing the relative abundances of each gene.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_proteomic</strong> – the path to to 2 column .csv file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">pandas DataFrame containing gene identifiers and relative abundances</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-Lipid"></span><div class="section" id="lipid">
<h2>Lipid<a class="headerlink" href="#lipid" title="Permalink to this headline">¶</a></h2>
<p>This module generates BOFsc for the lipid content of the cell.</p>
<dl class="function">
<dt id="Lipid.get_coefficients">
<code class="descclassname">Lipid.</code><code class="descname">get_coefficients</code><span class="sig-paren">(</span><em>path_to_lipidomic</em>, <em>path_to_bigg_dict</em>, <em>path_to_model</em>, <em>CELL_WEIGHT=280</em>, <em>LIPID_RATIO=0.091</em>, <em>R_WEIGHT=284.486</em><span class="sig-paren">)</span><a class="headerlink" href="#Lipid.get_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary of metabolite:coefficients for the lipid content of the cell. Lipids vary from a specie to another.
The lipidomic data provides the relative abundance of each lipid specie while the to_bigg_dict allows to convert identifiers given in the lipidomic data to BiGG identifiers for which
the metabolite weight is known and can be added easily to the biomass.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>lipidomic</strong> – a dictionary or dataframe of metabolites identified in the lipidomic experiment</li>
<li><strong>to_bigg_dict</strong> – a dictionary converting from the name present in the lipidomic data to BiGG identifiers. This dictionary is generated through manual curation from the modeller.</li>
<li><strong>CELL_WEIGHT</strong> – measured cell weight in femtograms, otherwise default</li>
<li><strong>LIPID_RATIO</strong> – measured lipid ratio of the cell, otherwise default</li>
<li><strong>R_WEIGHT</strong> – weight of a carbon chain, otherwise default. If the weight of the lipid is not known it will be inferred based on the number of R chains and this given weight.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary of metabolites and coefficients that can be used to update the biomass objective function.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Lipid.update_biomass_coefficients">
<code class="descclassname">Lipid.</code><code class="descname">update_biomass_coefficients</code><span class="sig-paren">(</span><em>dict_of_coefficients</em>, <em>model</em><span class="sig-paren">)</span><a class="headerlink" href="#Lipid.update_biomass_coefficients" title="Permalink to this definition">¶</a></dt>
<dd><p>Updates the biomass coefficients given the input metabolite:coefficient dictionary.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>dict_of_coefficients</strong> – dictionary of metabolites and coefficients</li>
<li><strong>model</strong> – model to update</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">The biomass objective function is updated.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Lipid._make_compliant_lipidomic">
<code class="descclassname">Lipid.</code><code class="descname">_make_compliant_lipidomic</code><span class="sig-paren">(</span><em>path_to_lipidomic</em><span class="sig-paren">)</span><a class="headerlink" href="#Lipid._make_compliant_lipidomic" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the 2 column .csv file that connects the lipid annotation from your lipid abundances file to identifiers in the model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_lipidomic</strong> – </td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a pandas dataframe containing lipid name and relative abundance</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Lipid._make_compliant_bigg">
<code class="descclassname">Lipid.</code><code class="descname">_make_compliant_bigg</code><span class="sig-paren">(</span><em>path_to_bigg_dict</em><span class="sig-paren">)</span><a class="headerlink" href="#Lipid._make_compliant_bigg" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the 2 column .csv file that connects the lipid annotation from your lipid abundances file to identifiers in the model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_bigg_dict</strong> – the path to the convertion 2 column .csv file</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a dictionary for convertion</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="Lipid._import_model">
<code class="descclassname">Lipid.</code><code class="descname">_import_model</code><span class="sig-paren">(</span><em>path_to_model</em><span class="sig-paren">)</span><a class="headerlink" href="#Lipid._import_model" title="Permalink to this definition">¶</a></dt>
<dd><p>This private function imports the model in either .json or .xml format</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path_to_model</strong> – the path to the model. Supported format are .json and .xml</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">model</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<span class="target" id="module-GAM"></span><div class="section" id="gam">
<h2>GAM<a class="headerlink" href="#gam" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="GAM.experimental_GAM">
<code class="descclassname">GAM.</code><code class="descname">experimental_GAM</code><span class="sig-paren">(</span><em>path_to_data</em>, <em>path_to_model</em>, <em>show_GAM=False</em><span class="sig-paren">)</span><a class="headerlink" href="#GAM.experimental_GAM" title="Permalink to this definition">¶</a></dt>
<dd><p>Description</p>
<p>Growth-associated maintenance (GAM) is the ATP cost of assembling macromolecules in the organism.
This function calculates GAM from provided path to experimental data. This data includes growth rates on
different carbon sources, the associated uptake rate for each carbon source and the secretion rates of metabolic
wastes. More information on the format in which to provide the experimental data is available on github.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path_to_data</strong> – The data file is the outcome of the HPLC growth, uptake and secretion rate experiment.</li>
<li><strong>path_to_model</strong> – The path to the model, json or sbml formats supported</li>
<li><strong>show_GAM</strong> – bool, will associated colors with carbon sources for easier display later</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="license.html" class="btn btn-neutral float-right" title="License" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="NGAM.html" class="btn btn-neutral" title="NGAM" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Jean-Christophe Lachance.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.0.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>