{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Make sure dependencies are installed\n",
    "import cobra\n",
    "import BOFdat as bd\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sb\n",
    "from cobra.util.solver import linear_reaction_coefficients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sb.set_style('whitegrid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Give the path to each file as function parameters\n",
    "#Genome file in BioPython supported format (.faa, .fna) and GenBank file \n",
    "#also in BioPython supported format (.gb, .gbff)\n",
    "genome = 'Ecoli_DNA.fna'\n",
    "genbank = 'Ecoli_K12_MG1655.gbff'\n",
    "\n",
    "#OMICs data as a 2 column csv file, gene and abundance\n",
    "transcriptomic = 'transcriptomic.csv'\n",
    "proteomic = 'proteomic.csv'\n",
    "\n",
    "#Lipidomic abundances and conversion to model identifier\n",
    "lipidomic_abundances = 'lipidomic_abundances.csv'\n",
    "lipidomic_conversion = 'lipidomic_conversion.csv'\n",
    "\n",
    "#Metabolomic abundances and conversion to model identifier\n",
    "metabolomic_abundances = 'metabolomic_abundances.csv'\n",
    "metabolomic_conversion = 'metabolomic_conversion.csv'\n",
    "\n",
    "#Growth data on different carbon sources, uptake and secretion rates\n",
    "maintenance = 'maintenance.csv'\n",
    "\n",
    "#The model for which the coefficients are generated\n",
    "model = 'iML1515.json'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ecoli = cobra.io.load_json_model(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'data_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-5fc57b2d36ca>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     29\u001b[0m         \u001b[0mmean_error\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmean\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 31\u001b[0;31m     \u001b[0mdata_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mratio\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmean_error\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0mdata_df\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'data_df' is not defined"
     ]
    }
   ],
   "source": [
    "#calculate and plot fold change\n",
    "old_biomass = {}\n",
    "for m in ecoli.reactions.Ec_biomass_iML1515_core_75p37M.reactants:\n",
    "    old_biomass[m.id] = abs(ecoli.reactions.Ec_biomass_iML1515_core_75p37M.get_coefficient(m))\n",
    "\n",
    "mean_error = []\n",
    "mega_df = pd.DataFrame(index=np.linspace(0.01,0.9,num=50),columns=['dna','rna','protein','lipid'])\n",
    "for ratio in np.linspace(0.01,0.9,num=50):\n",
    "    dna_coeff = bd.dna.generate_coefficients(genome,model,DNA_RATIO=ratio)\n",
    "    rna_coeff = bd.rna.generate_coefficients(genbank,model,transcriptomic,TOTAL_RNA_RATIO=ratio)\n",
    "    prot_coeff = bd.protein.generate_coefficients(genbank,model,proteomic,PROTEIN_RATIO=ratio)\n",
    "    lipid_coeff = bd.lipid.generate_coefficients(lipidomic_abundances,lipidomic_conversion,model,LIPID_RATIO=ratio)\n",
    "    #metab_coeff = bd.metabolite.generate_coefficients_from_experimental_data(metabolomic_abundances,metabolomic_conversion,model,METAB_RATIO=ratio)\n",
    "    new_biomass = [dna_coeff, rna_coeff, prot_coeff, lipid_coeff]\n",
    "    #Generate a new biomass objective function by adding all the components together\n",
    "    mean_error = []\n",
    "    for i in range(len(new_biomass)):\n",
    "        d = new_biomass[i]\n",
    "        df1 = pd.DataFrame({'new_coefficient':d.values(),'metab':[m.id for m in d.keys()]})\n",
    "        df2 = pd.DataFrame({'original_coefficient':old_biomass.values(),'metab': old_biomass.keys()})\n",
    "        df3 = pd.merge(left=df1,right=df2,how='outer',on='metab')\n",
    "        #Remove data that is absurdely divergent\n",
    "        cut_off = (df3.new_coefficient.median() + df3.original_coefficient.median())/2 + (df3.new_coefficient.std()+df3.original_coefficient.std())\n",
    "        \n",
    "        #Get all data below cut-off\n",
    "        df3['difference'] = abs(df3.new_coefficient[df3['new_coefficient']<cut_off]) - abs(df3.original_coefficient[df3['original_coefficient']<cut_off])\n",
    "        df3['percent_error'] = abs(df3['difference']/df3.original_coefficient[df3['original_coefficient']<cut_off])*100\n",
    "        data = df3[df3['percent_error'].isnull() == False].percent_error\n",
    "        mean_error.append(data.mean())\n",
    "    \n",
    "    data_df.loc[ratio] = mean_error\n",
    "    \n",
    "data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot\n",
    "ax = mega_df.plot()\n",
    "#Plot the minimum value for each of them\n",
    "plt.plot()\n",
    "# zoom-in / limit the view to different portions of the data\n",
    "ax.set_ylim(0., 100.)  # outliers only\n",
    "plt.ylabel('Percent error')\n",
    "plt.xlabel('Weight fraction')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_error, weight_fraction_min = [],[]\n",
    "for col in mega_df.columns:\n",
    "    min_error.append(mega_df[col].min())\n",
    "    weight_fraction_min.append(mega_df[col].idxmin())\n",
    "result_df = pd.DataFrame({'min_error':min_error,'weight_fraction_min':weight_fraction_min},index=['dna','rna','protein','lipid'])\n",
    "result_df.plot(kind='bar')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>min_error</th>\n",
       "      <th>weight_fraction_min</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dna</th>\n",
       "      <td>7.081532</td>\n",
       "      <td>0.054949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rna</th>\n",
       "      <td>14.431379</td>\n",
       "      <td>0.306667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein</th>\n",
       "      <td>24.710127</td>\n",
       "      <td>0.630303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lipid</th>\n",
       "      <td>22.363994</td>\n",
       "      <td>0.216768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         min_error  weight_fraction_min\n",
       "dna       7.081532             0.054949\n",
       "rna      14.431379             0.306667\n",
       "protein  24.710127             0.630303\n",
       "lipid    22.363994             0.216768"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.63030303030303036"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mega_df.protein.idxmin()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mega_df.to_csv('results_fig.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dna</th>\n",
       "      <th>rna</th>\n",
       "      <th>protein</th>\n",
       "      <th>lipid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0.010000</th>\n",
       "      <td>80.5127</td>\n",
       "      <td>96.7831</td>\n",
       "      <td>98.5779</td>\n",
       "      <td>96.3309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.018990</th>\n",
       "      <td>62.9939</td>\n",
       "      <td>93.8911</td>\n",
       "      <td>97.2994</td>\n",
       "      <td>93.0323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.027980</th>\n",
       "      <td>45.475</td>\n",
       "      <td>90.9991</td>\n",
       "      <td>96.0209</td>\n",
       "      <td>89.7338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.036970</th>\n",
       "      <td>27.9562</td>\n",
       "      <td>88.1071</td>\n",
       "      <td>94.7424</td>\n",
       "      <td>86.4353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.045960</th>\n",
       "      <td>10.4373</td>\n",
       "      <td>85.2151</td>\n",
       "      <td>93.4639</td>\n",
       "      <td>83.1368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.054949</th>\n",
       "      <td>7.08153</td>\n",
       "      <td>82.3231</td>\n",
       "      <td>92.1854</td>\n",
       "      <td>79.8383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.063939</th>\n",
       "      <td>24.6004</td>\n",
       "      <td>79.4311</td>\n",
       "      <td>90.9069</td>\n",
       "      <td>76.5397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.072929</th>\n",
       "      <td>42.1192</td>\n",
       "      <td>76.5391</td>\n",
       "      <td>89.6284</td>\n",
       "      <td>73.2412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.081919</th>\n",
       "      <td>59.6381</td>\n",
       "      <td>73.6471</td>\n",
       "      <td>88.3499</td>\n",
       "      <td>69.9427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.090909</th>\n",
       "      <td>77.1569</td>\n",
       "      <td>70.7551</td>\n",
       "      <td>87.0714</td>\n",
       "      <td>66.6442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.099899</th>\n",
       "      <td>94.6758</td>\n",
       "      <td>67.8631</td>\n",
       "      <td>85.7929</td>\n",
       "      <td>63.3457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.108889</th>\n",
       "      <td>112.195</td>\n",
       "      <td>64.9712</td>\n",
       "      <td>84.5144</td>\n",
       "      <td>60.0471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.117879</th>\n",
       "      <td>129.714</td>\n",
       "      <td>62.0792</td>\n",
       "      <td>83.2359</td>\n",
       "      <td>56.7486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.126869</th>\n",
       "      <td>147.232</td>\n",
       "      <td>59.1872</td>\n",
       "      <td>81.9574</td>\n",
       "      <td>53.4501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.135859</th>\n",
       "      <td>164.751</td>\n",
       "      <td>56.2952</td>\n",
       "      <td>80.6789</td>\n",
       "      <td>50.1516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.144848</th>\n",
       "      <td>182.27</td>\n",
       "      <td>53.4032</td>\n",
       "      <td>79.4004</td>\n",
       "      <td>46.8531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.153838</th>\n",
       "      <td>199.789</td>\n",
       "      <td>50.5112</td>\n",
       "      <td>78.1219</td>\n",
       "      <td>43.5545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.162828</th>\n",
       "      <td>217.308</td>\n",
       "      <td>47.6192</td>\n",
       "      <td>76.8434</td>\n",
       "      <td>40.256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.171818</th>\n",
       "      <td>234.827</td>\n",
       "      <td>44.7272</td>\n",
       "      <td>75.5649</td>\n",
       "      <td>36.9575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.180808</th>\n",
       "      <td>252.345</td>\n",
       "      <td>41.8352</td>\n",
       "      <td>74.2864</td>\n",
       "      <td>33.659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.189798</th>\n",
       "      <td>269.864</td>\n",
       "      <td>38.9432</td>\n",
       "      <td>73.0079</td>\n",
       "      <td>30.3605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.198788</th>\n",
       "      <td>287.383</td>\n",
       "      <td>36.0512</td>\n",
       "      <td>71.7294</td>\n",
       "      <td>27.0619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.207778</th>\n",
       "      <td>304.902</td>\n",
       "      <td>33.1592</td>\n",
       "      <td>70.4509</td>\n",
       "      <td>23.7634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.216768</th>\n",
       "      <td>322.421</td>\n",
       "      <td>30.2672</td>\n",
       "      <td>69.1724</td>\n",
       "      <td>22.364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.225758</th>\n",
       "      <td>339.94</td>\n",
       "      <td>27.3753</td>\n",
       "      <td>67.8939</td>\n",
       "      <td>23.2915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.234747</th>\n",
       "      <td>357.459</td>\n",
       "      <td>24.4833</td>\n",
       "      <td>66.6154</td>\n",
       "      <td>24.219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.243737</th>\n",
       "      <td>374.977</td>\n",
       "      <td>21.5913</td>\n",
       "      <td>65.3369</td>\n",
       "      <td>25.1465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.252727</th>\n",
       "      <td>392.496</td>\n",
       "      <td>18.6993</td>\n",
       "      <td>64.0584</td>\n",
       "      <td>26.074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.261717</th>\n",
       "      <td>410.015</td>\n",
       "      <td>16.9694</td>\n",
       "      <td>62.7799</td>\n",
       "      <td>27.0014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.270707</th>\n",
       "      <td>427.534</td>\n",
       "      <td>16.4073</td>\n",
       "      <td>61.5014</td>\n",
       "      <td>27.9289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.639293</th>\n",
       "      <td>1145.81</td>\n",
       "      <td>105.656</td>\n",
       "      <td>24.7107</td>\n",
       "      <td>134.566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.648283</th>\n",
       "      <td>1163.33</td>\n",
       "      <td>108.548</td>\n",
       "      <td>24.9176</td>\n",
       "      <td>137.864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.657273</th>\n",
       "      <td>1180.84</td>\n",
       "      <td>111.44</td>\n",
       "      <td>25.1244</td>\n",
       "      <td>141.163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.666263</th>\n",
       "      <td>1198.36</td>\n",
       "      <td>114.332</td>\n",
       "      <td>25.3313</td>\n",
       "      <td>144.461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.675253</th>\n",
       "      <td>1215.88</td>\n",
       "      <td>117.224</td>\n",
       "      <td>25.5382</td>\n",
       "      <td>147.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.684242</th>\n",
       "      <td>1233.4</td>\n",
       "      <td>120.116</td>\n",
       "      <td>25.745</td>\n",
       "      <td>151.058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.693232</th>\n",
       "      <td>1250.92</td>\n",
       "      <td>123.008</td>\n",
       "      <td>25.9519</td>\n",
       "      <td>154.357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.702222</th>\n",
       "      <td>1268.44</td>\n",
       "      <td>125.9</td>\n",
       "      <td>26.2436</td>\n",
       "      <td>157.655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.711212</th>\n",
       "      <td>1285.96</td>\n",
       "      <td>128.792</td>\n",
       "      <td>26.5795</td>\n",
       "      <td>160.954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.720202</th>\n",
       "      <td>1303.48</td>\n",
       "      <td>131.684</td>\n",
       "      <td>26.9155</td>\n",
       "      <td>164.252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.729192</th>\n",
       "      <td>1321</td>\n",
       "      <td>134.576</td>\n",
       "      <td>27.2515</td>\n",
       "      <td>167.551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.738182</th>\n",
       "      <td>1338.51</td>\n",
       "      <td>137.468</td>\n",
       "      <td>27.5875</td>\n",
       "      <td>170.849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.747172</th>\n",
       "      <td>1356.03</td>\n",
       "      <td>140.36</td>\n",
       "      <td>27.9484</td>\n",
       "      <td>174.148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.756162</th>\n",
       "      <td>1373.55</td>\n",
       "      <td>143.252</td>\n",
       "      <td>28.405</td>\n",
       "      <td>177.446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.765152</th>\n",
       "      <td>1391.07</td>\n",
       "      <td>146.144</td>\n",
       "      <td>28.8616</td>\n",
       "      <td>180.745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.774141</th>\n",
       "      <td>1408.59</td>\n",
       "      <td>149.036</td>\n",
       "      <td>29.3181</td>\n",
       "      <td>184.043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.783131</th>\n",
       "      <td>1426.11</td>\n",
       "      <td>151.928</td>\n",
       "      <td>29.7747</td>\n",
       "      <td>187.342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.792121</th>\n",
       "      <td>1443.63</td>\n",
       "      <td>154.82</td>\n",
       "      <td>30.3366</td>\n",
       "      <td>190.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.801111</th>\n",
       "      <td>1461.15</td>\n",
       "      <td>157.712</td>\n",
       "      <td>30.9079</td>\n",
       "      <td>193.939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.810101</th>\n",
       "      <td>1478.67</td>\n",
       "      <td>160.604</td>\n",
       "      <td>31.4792</td>\n",
       "      <td>197.237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.819091</th>\n",
       "      <td>1496.18</td>\n",
       "      <td>163.496</td>\n",
       "      <td>32.0588</td>\n",
       "      <td>200.536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.828081</th>\n",
       "      <td>1513.7</td>\n",
       "      <td>166.388</td>\n",
       "      <td>32.7399</td>\n",
       "      <td>203.834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.837071</th>\n",
       "      <td>1531.22</td>\n",
       "      <td>169.28</td>\n",
       "      <td>33.421</td>\n",
       "      <td>207.133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.846061</th>\n",
       "      <td>1548.74</td>\n",
       "      <td>172.172</td>\n",
       "      <td>34.1022</td>\n",
       "      <td>210.431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.855051</th>\n",
       "      <td>1566.26</td>\n",
       "      <td>175.064</td>\n",
       "      <td>34.7833</td>\n",
       "      <td>213.73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.864040</th>\n",
       "      <td>1583.78</td>\n",
       "      <td>177.956</td>\n",
       "      <td>35.4644</td>\n",
       "      <td>217.029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.873030</th>\n",
       "      <td>1601.3</td>\n",
       "      <td>180.848</td>\n",
       "      <td>36.1455</td>\n",
       "      <td>220.327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.882020</th>\n",
       "      <td>1618.82</td>\n",
       "      <td>183.74</td>\n",
       "      <td>36.8267</td>\n",
       "      <td>223.626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.891010</th>\n",
       "      <td>1636.33</td>\n",
       "      <td>186.632</td>\n",
       "      <td>37.5078</td>\n",
       "      <td>226.924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0.900000</th>\n",
       "      <td>1653.85</td>\n",
       "      <td>189.524</td>\n",
       "      <td>38.2286</td>\n",
       "      <td>230.223</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              dna      rna  protein    lipid\n",
       "0.010000  80.5127  96.7831  98.5779  96.3309\n",
       "0.018990  62.9939  93.8911  97.2994  93.0323\n",
       "0.027980   45.475  90.9991  96.0209  89.7338\n",
       "0.036970  27.9562  88.1071  94.7424  86.4353\n",
       "0.045960  10.4373  85.2151  93.4639  83.1368\n",
       "0.054949  7.08153  82.3231  92.1854  79.8383\n",
       "0.063939  24.6004  79.4311  90.9069  76.5397\n",
       "0.072929  42.1192  76.5391  89.6284  73.2412\n",
       "0.081919  59.6381  73.6471  88.3499  69.9427\n",
       "0.090909  77.1569  70.7551  87.0714  66.6442\n",
       "0.099899  94.6758  67.8631  85.7929  63.3457\n",
       "0.108889  112.195  64.9712  84.5144  60.0471\n",
       "0.117879  129.714  62.0792  83.2359  56.7486\n",
       "0.126869  147.232  59.1872  81.9574  53.4501\n",
       "0.135859  164.751  56.2952  80.6789  50.1516\n",
       "0.144848   182.27  53.4032  79.4004  46.8531\n",
       "0.153838  199.789  50.5112  78.1219  43.5545\n",
       "0.162828  217.308  47.6192  76.8434   40.256\n",
       "0.171818  234.827  44.7272  75.5649  36.9575\n",
       "0.180808  252.345  41.8352  74.2864   33.659\n",
       "0.189798  269.864  38.9432  73.0079  30.3605\n",
       "0.198788  287.383  36.0512  71.7294  27.0619\n",
       "0.207778  304.902  33.1592  70.4509  23.7634\n",
       "0.216768  322.421  30.2672  69.1724   22.364\n",
       "0.225758   339.94  27.3753  67.8939  23.2915\n",
       "0.234747  357.459  24.4833  66.6154   24.219\n",
       "0.243737  374.977  21.5913  65.3369  25.1465\n",
       "0.252727  392.496  18.6993  64.0584   26.074\n",
       "0.261717  410.015  16.9694  62.7799  27.0014\n",
       "0.270707  427.534  16.4073  61.5014  27.9289\n",
       "...           ...      ...      ...      ...\n",
       "0.639293  1145.81  105.656  24.7107  134.566\n",
       "0.648283  1163.33  108.548  24.9176  137.864\n",
       "0.657273  1180.84   111.44  25.1244  141.163\n",
       "0.666263  1198.36  114.332  25.3313  144.461\n",
       "0.675253  1215.88  117.224  25.5382   147.76\n",
       "0.684242   1233.4  120.116   25.745  151.058\n",
       "0.693232  1250.92  123.008  25.9519  154.357\n",
       "0.702222  1268.44    125.9  26.2436  157.655\n",
       "0.711212  1285.96  128.792  26.5795  160.954\n",
       "0.720202  1303.48  131.684  26.9155  164.252\n",
       "0.729192     1321  134.576  27.2515  167.551\n",
       "0.738182  1338.51  137.468  27.5875  170.849\n",
       "0.747172  1356.03   140.36  27.9484  174.148\n",
       "0.756162  1373.55  143.252   28.405  177.446\n",
       "0.765152  1391.07  146.144  28.8616  180.745\n",
       "0.774141  1408.59  149.036  29.3181  184.043\n",
       "0.783131  1426.11  151.928  29.7747  187.342\n",
       "0.792121  1443.63   154.82  30.3366   190.64\n",
       "0.801111  1461.15  157.712  30.9079  193.939\n",
       "0.810101  1478.67  160.604  31.4792  197.237\n",
       "0.819091  1496.18  163.496  32.0588  200.536\n",
       "0.828081   1513.7  166.388  32.7399  203.834\n",
       "0.837071  1531.22   169.28   33.421  207.133\n",
       "0.846061  1548.74  172.172  34.1022  210.431\n",
       "0.855051  1566.26  175.064  34.7833   213.73\n",
       "0.864040  1583.78  177.956  35.4644  217.029\n",
       "0.873030   1601.3  180.848  36.1455  220.327\n",
       "0.882020  1618.82   183.74  36.8267  223.626\n",
       "0.891010  1636.33  186.632  37.5078  226.924\n",
       "0.900000  1653.85  189.524  38.2286  230.223\n",
       "\n",
       "[100 rows x 4 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{<Metabolite dgtp_c at 0x7fd25beffc90>: 0.027017,\n",
       " <Metabolite fad_c at 0x7fd25bf24410>: 0.000223,\n",
       " <Metabolite succoa_c at 0x7fd25bf35350>: 9.8e-05,\n",
       " <Metabolite his__L_c at 0x7fd25bf36290>: 0.094738,\n",
       " <Metabolite ribflv_c at 0x7fd25bf362d0>: 0.000223,\n",
       " <Metabolite so4_c at 0x7fd25bf36350>: 0.004338,\n",
       " <Metabolite pe161_p at 0x7fd25bf36dd0>: 0.075214,\n",
       " <Metabolite 2fe2s_c at 0x7fd25bf36f50>: 2.6e-05,\n",
       " <Metabolite gtp_c at 0x7fd25bf48150>: 0.215096,\n",
       " <Metabolite btn_c at 0x7fd25bf48290>: 2e-06,\n",
       " <Metabolite val__L_c at 0x7fd25bf488d0>: 0.423162,\n",
       " <Metabolite pro__L_c at 0x7fd25bf5b250>: 0.221055,\n",
       " <Metabolite coa_c at 0x7fd25bf6d050>: 0.000576,\n",
       " <Metabolite asn__L_c at 0x7fd25bf6d090>: 0.241055,\n",
       " <Metabolite leu__L_c at 0x7fd25bf6d150>: 0.450531,\n",
       " <Metabolite 4fe4s_c at 0x7fd25bf6dc90>: 0.00026,\n",
       " <Metabolite cl_c at 0x7fd25bf7f390>: 0.005205,\n",
       " <Metabolite murein5px4p_p at 0x7fd25bf92190>: 0.013894,\n",
       " <Metabolite pheme_c at 0x7fd25bf92cd0>: 0.000223,\n",
       " <Metabolite 2ohph_c at 0x7fd25bf92d90>: 0.000223,\n",
       " <Metabolite mobd_c at 0x7fd25bfa4090>: 7e-06,\n",
       " <Metabolite fe3_c at 0x7fd25bfc9090>: 0.007808,\n",
       " <Metabolite udcpdp_c at 0x7fd25bfc9750>: 5.5e-05,\n",
       " <Metabolite ala__L_c at 0x7fd25bfdb5d0>: 0.513689,\n",
       " <Metabolite phe__L_c at 0x7fd25bfdb650>: 0.185265,\n",
       " <Metabolite mlthf_c at 0x7fd25bfdb710>: 0.000223,\n",
       " <Metabolite kdo2lipid4_e at 0x7fd25bfdbe10>: 0.019456,\n",
       " <Metabolite h2o_c at 0x7fd25bfdbed0>: 70.028756,\n",
       " <Metabolite gly_c at 0x7fd25bfee090>: 0.612638,\n",
       " <Metabolite cu2_c at 0x7fd25bfee490>: 0.000709,\n",
       " <Metabolite ca2_c at 0x7fd25bfeec10>: 0.005205,\n",
       " <Metabolite datp_c at 0x7fd25c000390>: 0.026166,\n",
       " <Metabolite cobalt2_c at 0x7fd25c012090>: 2.5e-05,\n",
       " <Metabolite fe2_c at 0x7fd25c012510>: 0.006715,\n",
       " <Metabolite nadp_c at 0x7fd25c025210>: 0.000447,\n",
       " <Metabolite glu__L_c at 0x7fd25c025250>: 0.26316,\n",
       " <Metabolite ile__L_c at 0x7fd25c025310>: 0.290529,\n",
       " <Metabolite gln__L_c at 0x7fd25c025950>: 0.26316,\n",
       " <Metabolite zn2_c at 0x7fd25c038410>: 0.000341,\n",
       " <Metabolite tyr__L_c at 0x7fd25c038610>: 0.137896,\n",
       " <Metabolite cys__L_c at 0x7fd25c038650>: 0.09158,\n",
       " <Metabolite atp_c at 0x7fd25c038d10>: 75.55223,\n",
       " <Metabolite thf_c at 0x7fd25c038e50>: 0.000223,\n",
       " <Metabolite pe160_p at 0x7fd25c049490>: 0.063814,\n",
       " <Metabolite dttp_c at 0x7fd25c049f10>: 0.026166,\n",
       " <Metabolite trp__L_c at 0x7fd25c05cd90>: 0.056843,\n",
       " <Metabolite lys__L_c at 0x7fd25c05ce10>: 0.343161,\n",
       " <Metabolite dctp_c at 0x7fd25c06e750>: 0.027017,\n",
       " <Metabolite utp_c at 0x7fd25c06ed50>: 0.144104,\n",
       " <Metabolite k_c at 0x7fd25c06ef90>: 0.195193,\n",
       " <Metabolite asp__L_c at 0x7fd25c080390>: 0.241055,\n",
       " <Metabolite met__L_c at 0x7fd25c080c50>: 0.153686,\n",
       " <Metabolite amet_c at 0x7fd25c093450>: 0.000223,\n",
       " <Metabolite nh4_c at 0x7fd25c0a5f90>: 0.013013,\n",
       " <Metabolite thmpp_c at 0x7fd25c0b87d0>: 0.000223,\n",
       " <Metabolite ni2_c at 0x7fd25c0b88d0>: 0.000323,\n",
       " <Metabolite ser__L_c at 0x7fd25c0b8c90>: 0.215792,\n",
       " <Metabolite sheme_c at 0x7fd25c0cac90>: 0.000223,\n",
       " <Metabolite ctp_c at 0x7fd25c0dca50>: 0.133508,\n",
       " <Metabolite arg__L_c at 0x7fd25c0ef110>: 0.295792,\n",
       " <Metabolite nad_c at 0x7fd25c0ef6d0>: 0.001831,\n",
       " <Metabolite mg2_c at 0x7fd25c0efa90>: 0.008675,\n",
       " <Metabolite mn2_c at 0x7fd25c126690>: 0.000691,\n",
       " <Metabolite 10fthf_c at 0x7fd25c126910>: 0.000223,\n",
       " <Metabolite pydx5p_c at 0x7fd25c126a90>: 0.000223,\n",
       " <Metabolite thr__L_c at 0x7fd25c126f10>: 0.253687}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_biomass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0025491558328838634"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3.new_coefficient.median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(u'5mta_c', 0.00041621734216333415, nan)\n",
      "(u'ump_p', 0.0075570898269380635, nan)\n",
      "(u'alltn_c', 0.00018634418894550607, nan)\n",
      "(u'acglu_c', 0.0005151385466297172, nan)\n",
      "(u'ru5p__D_c', 0.00046698581827805854, nan)\n",
      "(u'ptrc_e', 0.07602716531618797, nan)\n",
      "(u'cit_c', 0.0008966263382877431, nan)\n",
      "(u'udpglcur_c', 0.0004600208040941871, nan)\n",
      "(u'acon_C_c', 0.00034600570946104946, nan)\n",
      "(u'spmd_p', 0.0007415111234063207, nan)\n",
      "(u'ahcys_c', 7.376475975594234e-05, nan)\n",
      "(u'alaala_c', 0.002572334755730239, nan)\n",
      "(u'cbasp_c', 0.006913039729684365, nan)\n",
      "(u'ade_p', 0.00015285613203150461, nan)\n",
      "(u'uacgam_c', 0.0019174488689623086, nan)\n",
      "(u'quin_e', 0.0279834419726804, nan)\n",
      "(u'pser__L_e', 0.00018246741863738456, nan)\n",
      "(u'cdpdhdecg_c', 0.0001228001776139871, nan)\n",
      "(u'nadph_c', 0.00047612751073745414, nan)\n",
      "(u'gdpfuc_c', 0.0004032004990218537, nan)\n",
      "(u'g6p_c', 0.0009381372805471971, nan)\n",
      "(u'acgam1p_e', 2.6132246900691972e-05, nan)\n",
      "(u'adphep_DD_c', 0.001029871665236765, nan)\n",
      "(u'sl26da_c', 0.03814413584952316, nan)\n",
      "(u'pep_c', 0.0007732814444175598, nan)\n",
      "(u'dtdp_c', 0.0001383534276373939, nan)\n",
      "(u'dadp_c', 7.372367790728449e-05, nan)\n",
      "(u'15dap_e', 0.00098899007526709, nan)\n",
      "(u'acorn_c', 0.0016360002692223374, nan)\n",
      "(u'succoa_c', 0.0019594632443701407, 9.8e-05)\n",
      "(u'itp_c', 0.0012380970445704247, nan)\n",
      "(u'orn_c', 0.000569616908648829, nan)\n",
      "(u'imp_e', 8.819864133499483e-05, nan)\n",
      "(u'f6p_c', 0.0007315806043854583, nan)\n",
      "(u'r5p_e', 0.00015566323896163713, nan)\n",
      "(u'ametam_c', 0.0004548738341515636, nan)\n",
      "(u'6pgc_c', 0.0005659610330723644, nan)\n",
      "(u'pphn_c', 0.0014128138013305596, nan)\n",
      "(u'gmp_p', 0.0016433950405265463, nan)\n",
      "(u'fad_c', 4.514106719343422e-05, 0.000223)\n",
      "(u'ncam_c', 0.023477246682102125, nan)\n",
      "(u'dhap_c', 0.0005701900977181101, nan)\n",
      "(u'fdp_c', 0.008842000460160832, nan)\n",
      "(u'g1p_c', 0.000619979241935111, nan)\n",
      "(u'akg_p', 0.0016877869275670903, nan)\n",
      "(u'sbt6p_c', 0.0014477828750169206, nan)\n",
      "(u'gam6p_p', 0.0005276917465741692, nan)\n",
      "(u'orot_c', 0.0054529810998160105, nan)\n",
      "(u'agm_e', 0.0002653499545727775, nan)\n",
      "(u'gthrd_e', 0.00014865368532629956, nan)\n",
      "(u'fum_e', 0.0037980766601022, nan)\n",
      "(u'nadh_c', 0.0012634518642535056, nan)\n",
      "(u'ap4a_c', 1.2300911210146102e-05, nan)\n",
      "(u'hom__L_p', 0.0015228555819769452, nan)\n",
      "(u'mal__L_e', 0.025387653368938754, nan)\n",
      "(u'dcmp_c', 8.836637994627787e-06, nan)\n",
      "(u'dtmp_c', 0.0020309737019075695, nan)\n",
      "(u's7p_c', 0.0004058673018297641, nan)\n",
      "(u'gthox_c', 0.033715277430682415, nan)\n",
      "(u'3pg_c', 0.003644859411832132, nan)\n",
      "(u'amp_c', 0.00018757502815699702, nan)\n",
      "(u'pyam5p_c', 0.0001451521500499204, nan)\n",
      "(u'acgam6p_c', 2.7323699238762607e-05, nan)\n",
      "(u'histd_c', 0.0005355389580039041, nan)\n",
      "(u'2ippm_c', 0.0012215637289384926, nan)\n",
      "(u'dcdp_c', 4.8308535134163353e-05, nan)\n",
      "(u'dhor__S_c', 0.0003096543995519533, nan)\n",
      "(u'citr__L_c', 0.0012041876506304287, nan)\n",
      "(u'coa_c', nan, 0.000576)\n",
      "(u'zn2_c', nan, 0.000341)\n",
      "(u'gly_c', nan, 0.612638)\n",
      "(u'val__L_c', nan, 0.423162)\n",
      "(u'h2o_c', nan, 70.028756)\n",
      "(u'arg__L_c', nan, 0.295792)\n",
      "(u'dgtp_c', nan, 0.027017)\n",
      "(u'leu__L_c', nan, 0.450531)\n",
      "(u'udcpdp_c', nan, 5.5e-05)\n",
      "(u'mg2_c', nan, 0.008675)\n",
      "(u'murein5px4p_p', nan, 0.013894)\n",
      "(u'met__L_c', nan, 0.153686)\n",
      "(u'nadp_c', nan, 0.000447)\n",
      "(u'glu__L_c', nan, 0.26316)\n",
      "(u'cobalt2_c', nan, 2.5e-05)\n",
      "(u'ctp_c', nan, 0.133508)\n",
      "(u'dttp_c', nan, 0.026166)\n",
      "(u'gln__L_c', nan, 0.26316)\n",
      "(u'his__L_c', nan, 0.094738)\n",
      "(u'cys__L_c', nan, 0.09158)\n",
      "(u'cl_c', nan, 0.005205)\n",
      "(u'pe161_p', nan, 0.075214)\n",
      "(u'ile__L_c', nan, 0.290529)\n",
      "(u'mobd_c', nan, 7e-06)\n",
      "(u'pro__L_c', nan, 0.221055)\n",
      "(u'4fe4s_c', nan, 0.00026)\n",
      "(u'asp__L_c', nan, 0.241055)\n",
      "(u'ribflv_c', nan, 0.000223)\n",
      "(u'2fe2s_c', nan, 2.6e-05)\n",
      "(u'ca2_c', nan, 0.005205)\n",
      "(u'amet_c', nan, 0.000223)\n",
      "(u'lys__L_c', nan, 0.343161)\n",
      "(u'cu2_c', nan, 0.000709)\n",
      "(u'atp_c', nan, 75.55223)\n",
      "(u'pheme_c', nan, 0.000223)\n",
      "(u'ni2_c', nan, 0.000323)\n",
      "(u'fe2_c', nan, 0.006715)\n",
      "(u'tyr__L_c', nan, 0.137896)\n",
      "(u'utp_c', nan, 0.144104)\n",
      "(u'pydx5p_c', nan, 0.000223)\n",
      "(u'datp_c', nan, 0.026166)\n",
      "(u'2ohph_c', nan, 0.000223)\n",
      "(u'trp__L_c', nan, 0.056843)\n",
      "(u'fe3_c', nan, 0.007808)\n",
      "(u'ala__L_c', nan, 0.513689)\n",
      "(u'pe160_p', nan, 0.063814)\n",
      "(u'phe__L_c', nan, 0.185265)\n",
      "(u'kdo2lipid4_e', nan, 0.019456)\n",
      "(u'nh4_c', nan, 0.013013)\n",
      "(u'thf_c', nan, 0.000223)\n",
      "(u'10fthf_c', nan, 0.000223)\n",
      "(u'mn2_c', nan, 0.000691)\n",
      "(u'sheme_c', nan, 0.000223)\n",
      "(u'nad_c', nan, 0.001831)\n",
      "(u'so4_c', nan, 0.004338)\n",
      "(u'ser__L_c', nan, 0.215792)\n",
      "(u'mlthf_c', nan, 0.000223)\n",
      "(u'dctp_c', nan, 0.027017)\n",
      "(u'btn_c', nan, 2e-06)\n",
      "(u'k_c', nan, 0.195193)\n",
      "(u'thr__L_c', nan, 0.253687)\n",
      "(u'asn__L_c', nan, 0.241055)\n",
      "(u'thmpp_c', nan, 0.000223)\n",
      "(u'gtp_c', nan, 0.215096)\n"
     ]
    }
   ],
   "source": [
    "for i,row in df3.iterrows():\n",
    "    print(row.metab, row.new_coefficient, row.original_coefficient)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "labels = df3[df3['fold_change'].isnull() == False].metab\n",
    "x_pos = np.arange(len(labels))\n",
    "plt.bar(x_pos,data.sort_values())\n",
    "plt.show()\n",
    "\n",
    "plt.clf()\n",
    "plt.plot(np.linspace(0.1,0.9,num=10),mean_error,'bo')\n",
    "plt.xlabel('Protein ratio')\n",
    "plt.ylabel('Mean fold change')\n",
    "plt.show()\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
